### PR TITLE
get rid of registry

### PR DESCRIPTION
**Pull request by joelgrus**
as discussed, replace with `Registrable` mix-in. this fixes all of the circular reference issues.  this PR ended up cleaner than I expected. 

I know there was some unhappiness about the exact code structure the registry had forced us into, this doesn't address any of that. also the way I handled `Initializer`s is somewhat hacky, but I can't think of a better way that doesn't involve wrapper classes.

### CODE REVIEW

File: allennlp/common/registrable.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    After which you can call ``BaseClass.list_available()`` to get the keys for the
+    registered subclasses, and ``BaseClass.by_name(name)`` to get the corresponding subclass.
+    Note that the registry stores the subclasses themselves; not class instances.
+    In most cases you would then call ``from_params(params)`` on the returned subclass.
+
+    You can specify a default by setting ``BaseClass.default_implementation``.
+    If it is set, it will be the first element of ``list_available()``.
+    """
+
+    _registry = defaultdict(dict)  # type: Dict[type, Dict[str, type]]
```
Comment (matt-gardner) [inline allennlp/common/registrable.py]: I'm surprised this accepts lowercase `type` instead of `Type`.

File: allennlp/common/registrable.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+        keys = list(Registrable._registry[cls].keys())
+        default = cls.default_implementation
+
+        if default is None:
+            return keys
+        elif default not in keys:
+            message = "Default implementation %s is not registered" % default
+            raise ConfigurationError(message)
+        else:
+            return [default] + [k for k in keys if k != default]
```
Comment (matt-gardner) [inline allennlp/common/registrable.py]: Huh, yeah, that's way better than how I wrote it.

File: allennlp/data/data_iterator.py [hunk 3]
```diff
@@ -52,9 +55,8 @@ def _create_batches(self, dataset: Dataset, shuffle: bool) -> List[List[Instance
...
     @classmethod
     def from_params(cls, params: Params):
-        from allennlp.experiments.registry import Registry
         # TODO(Mark): The adaptive iterator will need a bit of work here,
         # to retrieve the scaling function etc.
 
-        iterator_type = params.pop_choice("type", Registry.list_data_iterators())
-        return Registry.get_data_iterator(iterator_type)(**params.as_dict())  # type: ignore
+        iterator_type = params.pop_choice("type", cls.list_available())
+        return cls.by_name(iterator_type)(**params.as_dict())  # type: ignore
```
Comment (matt-gardner) [inline allennlp/data/data_iterator.py]: Do we actually need the `# type: ignore` here?  We don't have it in other places.  Also, there is no listed return type for the method - maybe that's the reason there was a type failure?

File: allennlp/data/dataset_reader.py [hunk 4]
```diff
@@ -1,8 +1,11 @@
+from typing import List, Dict, TypeVar, Type, Generic  # pylint: disable=unused-import
```
Comment (matt-gardner) [inline allennlp/data/dataset_reader.py]: Why do we need these imports?

File: allennlp/data/dataset_reader.py [hunk 5]
```diff
@@ -16,6 +19,5 @@ def read(self, file_path: str) -> Dataset:
 
     @classmethod
     def from_params(cls, params: Params):
```
Comment (matt-gardner) [inline allennlp/data/dataset_reader.py]: Can you add a return type here, and anywhere else on these where it's missing one?

File: allennlp/modules/seq2seq_encoders/__init__.py [hunk 6]
```diff
@@ -4,26 +4,26 @@
 
 from allennlp.common import Params
 from allennlp.common.checks import ConfigurationError
-from allennlp.experiments import Registry
+from allennlp.modules import Seq2SeqEncoder
 from allennlp.modules.seq2seq_encoders.pytorch_seq2seq_wrapper import PytorchSeq2SeqWrapper
 
 class _Seq2SeqWrapper:
     """
-    The :class:`Registry` needs to have a ``Type[Seq2SeqEncoder]`` as the value registered for each
```
Comment (matt-gardner) [inline allennlp/modules/seq2seq_encoders/__init__.py]: @DeNeutoy, I'm guessing we can simplify this quite a bit, just handling this specially `Seq2SeqEncoder.from_params()`, now that we're not going through the registry.  We can definitely worry about that later, in another PR, though.

File: allennlp/common/registrable.py (hunk 1 above)
Comment (joelgrus) [inline allennlp/common/registrable.py]: for some reason I was convinced that it wouldn't typecheck with `Type`, but it seems to, so I'll change it

File: allennlp/data/data_iterator.py (hunk 3 above)
Comment (joelgrus) [inline allennlp/data/data_iterator.py]: without it you get a `error: Too many arguments for "DataIterator"`.

the base class doesn't have a constructor that takes **kwargs, so mypy complains.

File: allennlp/data/dataset_reader.py (hunk 4 above)
Comment (joelgrus) [inline allennlp/data/dataset_reader.py]: we don't

File: allennlp/data/data_iterator.py (hunk 3 above)
Comment (matt-gardner) [inline allennlp/data/data_iterator.py]: Oh, right, we hadn't yet established the `from_params()` pattern when we wrote this code.  Can you add in a TODO to switch this to initialize `from_params()`?

File: allennlp/data/dataset_reader.py (hunk 5 above)
Comment (joelgrus) [inline allennlp/data/dataset_reader.py]: done

**Comment by matt-gardner** [comment]
@DeNeutoy, it's fine with me to put, e.g., `similarity_function.py` inside of `similarity_functions/`, but I'd still put `from allennlp.modules.similarity_functions.similarity_function import SimilarityFunction` inside `allennlp/modules/__init__.py` - that is, let the type be importable from the base module.

File: allennlp/data/data_iterator.py (hunk 3 above)
Comment (joelgrus) [inline allennlp/data/data_iterator.py]: will do

**Pull request merged**